### DISCUSSION TITLE
c2a-aobcの環境構築について

### DISCUSSION THREAD

**Discussion by KotaToyoda** [ut-issl/c2a-aobc#190] [Q&A]
c2a-aobcの環境構築についてお尋ねしたく，ご連絡させていただきました．現在，c2a-aobc（https://github.com/ut-issl/c2a-aobc/tree/develop）の環境構築を，READMEを参考に下記の手順で行いました．

1. c2a-aobc/Examplesをコピーし、c2a-aobc-exampleというリポジトリを新しく作る
2. c2a-aobc-example内に、c2a-aobcをsrc-aobcという名前でsubmoduleとして取り込む
3. chipKitをインストールし，ボード定義ファイルを追加して，vMicroの設定を行う
4. c2a-aobc-example/C2A-AOBC-EXAMPLE.slnをビルドする

この手順を行うと写真のようなエラーが出てしまい，ビルドが失敗します．この問題の解決方法をご教授いただければ幸いです．
![image](https://github.com/ut-issl/c2a-aobc/assets/131437019/910a0ab9-53c6-438c-b039-8842da4f4a41)

**Comment by 200km** [comment]
ご質問ありがとうございます。  
`c2a-aobc-example/C2A-AOBC-EXAMPLE.slnをビルドする`というのは具体的にはどのような操作をされていますか？vMicroでのビルドがおこなわれているか、Visual Studioの通常のビルド操作が行われていないかなどが気になっています。

**Reply by KotaToyoda** [reply]
ご返答ありがとうございます．
現在，下記の手順でビルドを行っています．
・写真の①でArduino 1.6/1.8を選択し，IDEのlocationを写真の②で選択
・ボードを写真の③で選択
・ポートを写真の④で選択
・写真の⑤のVisual Microのビルドを行う
![ikari1](https://github.com/ut-issl/c2a-aobc/assets/131437019/410b7604-2d8d-47ed-8688-a8beefc49924)
![ikari2](https://github.com/ut-issl/c2a-aobc/assets/131437019/e28e94ae-ba6d-4b5d-91c5-3d2c51aee185)

**Reply by 200km** [reply]
ありがとうございます。この手順は問題なさそうです。  
ここに貼ってくださった画面に見えているエラーメッセージからは、DTWI.hが見つからないと読み取れます。  
chipKitのインストールは正しくできていることは確認されていますか？インストールに成功していれば、chipKitディレクトリ配下のどこかにDTWI.hなどが存在しているはずなので検索をかけてみてください。  

他にもvMicro設定で必要なものがあるかもしれませんので、私の方ももう少し確認してみます。

seki-hiroさんから質問されているバージョン情報についても重要ですので教えていただけますと幸いです。

**Reply by KotaToyoda** [reply]
現在新たにc2a-aobcのバージョンv7.0.0で試しているところです．submoduleしたc2a-coreのバージョンはv3.10.0です．

chipKitのインストールは正しくできています．また，DTWI.hはchipKIT\hardware\pic32\2.1.0\libraries\DTWI内にありました．ですがc2a-aobc/__vmディレクトリ配下のCompile.vmps.xmlファイルを確認したところ，chipKIT\hardware\pic32\2.1.0\librariesのパス指定が見当たらなかったため，DTWI.h，DTWI.c（とDSPI.h，DSPI.c）をchipKIT\hardware\pic32\2.1.0\cores/pic32内にコピーペーストしたところ，写真のようにコンパイルできたようです．しかし，無理矢理コピーしただけなので，正しい手順ではないかと思われます．
<img width="959" alt="スクリーンショット 2023-08-22 174543" src="https://github.com/ut-issl/c2a-aobc/assets/131437019/47548ca7-0c64-495f-b84a-6fcf6aa5f8a3">

**Reply by 200km** [reply]
ありがとうございます。v7.0.0はまだDraft段階で公式release前なので、できればv6系列で試してもらえると幸いです。（その後すぐに正式releaseしたので使ってくださって構いません）  
説明いただいた現象から、Include libraries関連の設定が必要なのに抜けてしまっているのだと思います。  
ISSLの設定を確認してREADMEに追記しようと思います。

**Reply by 200km** [reply]
@suzuki-toshihir0 昔鈴木くんが書いていた下記の現象に似ている気もするのですが、その時はどうやって解決したんでしたっけ？

https://github.com/ut-issl/c2a-aobc/pull/114#issuecomment-1621556242

**Reply by suzuki-toshihir0** [reply]
DTWI，DSPIについては結局原因がよくわかっていないです．．．他のところでエラーが起きているときに何故かこれらのファイルが無いと言われたりなどもあります．試行錯誤しているうちに解消されてしまって，どこが原因かは突き止められていないですね．．．

**Reply by 200km** [reply]
@KotaToyoda 
`chipKIT\hardware\pic32\2.1.0\libraries\DTWI` の絶対パスを教えていただけないでしょうか。ユーザー名など隠したいものがあれば、`USERNAME`などテキトウな文字列に置き換えていただいて構いません。下記のIncludeDirectories設定との違いを確認したいです。

https://github.com/ut-issl/c2a-aobc/blob/6ff71bdda8eb071dd7df0a775a119ee89280648a/Examples/C2A_AOBC_EXAMPLE.vcxproj#L80

**Reply by KotaToyoda** [reply]
絶対パスは
C:\Users\^^^^\AppData\Local\Arduino15\packages\chipKIT\hardware\pic32\2.1.0\libraries\DTWI
です。

**Reply by 200km** [reply]
ありがとうございます。vcsprojで指定しているものと同じですね。vcsprojの中でclone直後は`USERNAME`となっている部分が、ビルドしてウィンドウを閉じて保存するを選ぶと環境に合わせたものに変わると思うのですが、そのような挙動は確認できていますか？

あと、コピーペーストした先である`cores/pic32`も同じようにvcsproj内で指定されているものです。これはとりこまれているのに`DTWI`ディレクトリは取り込まれていないのは少し不思議ですね。

**Reply by KotaToyoda** [reply]
C2A_AOBC_EXAMPLE.vcsprojの`AdditionalIncludeDirectories`の部分ですが，ビルド後は
`<AdditionalIncludeDirectories>$(ProjectDir)..\..\..\..\AppData\Local\arduino15\packages\chipKIT\hardware\pic32\2.1.0\cores\pic32;$(ProjectDir)..\..\..\..\AppData\Local\arduino15\packages\chipKIT\hardware\pic32\2.1.0\cores\pic32\avr;$(ProjectDir)..\..\..\..\AppData\Local\arduino15\packages\chipKIT\hardware\pic32\2.1.0\cores\pic32\pps;$(ProjectDir)..\..\..\..\AppData\Local\arduino15\packages\chipKIT\hardware\pic32\2.1.0\cores\pic32\util;$(ProjectDir)..\..\..\..\AppData\Local\arduino15\packages\chipKIT\hardware\pic32\2.1.0\variants\MXL795;$(ProjectDir)..\..\..\..\..\..\\Users\\^^^^\\AppData\\Local\\arduino15\\packages\\chipKIT\\hardware\\pic32C:\\Users\\^^^^\\AppData\\Local\\arduino15\\packages\\chipKIT\\tools\\pic32-tools\\1.43-pic32gcc\\pic32mx\\include;%(AdditionalIncludeDirectories)</AdditionalIncludeDirectories>`
に変わっていました．DTWI，DSPIディレクトリの指定がなくなってしまっているようです．

**Reply by 200km** [reply]
ISSLでは保存後もDTWI、DSPIのディレクトリ指定は無くならないので、そこに何か原因がありそうですね。念のための確認ですが、
`C:\Users^^^^\AppData\Local\Arduino15` と書いてくださった部分は、実際には`C:\Users\^^^^\AppData\Local\Arduino15`のような感じでしょうか？

**Reply by KotaToyoda** [reply]
すみません．`C:\Users\^^^^\AppData\Local\Arduino15`が正しいです．

**Reply by 200km** [reply]
@KotaToyoda 環境構築後、Visual Studioで開く前にvcsprojファイルを手動で編集し、DTWI、DSPIのディレクトリをお使いの環境のものに合わせた上でVisual Studioで開いていただくとどのようになりますか？

**Reply by KotaToyoda** [reply]
vcsprojファイルのDTWI、DSPIのディレクトリ指定を私の環境に合わせて編集してからVisual Studioで開いても，ビルドはできませんでした．また，ウインドウを閉じて保存するとvcsprojファイルが書き換わってしまい，上でコメントしたようにDTWI、DSPIのディレクトリ指定が無くなってしまいました．

**Reply by 200km** [reply]
試していただきありがとうございます。  
ISSL側で設定している何かしらのものが手順やファイルの中から漏れてしまっているようです。確認してみますが、時間がかかってしまうかもしれません。

**Reply by KotaToyoda** [reply]
承知しました．確認のほど，よろしくお願いします．

**Reply by Sota2811** [reply]
@200km c2a-aobcのパスについてお尋ねしたいことがあります．相対パスがsrc_core, src_userから始まっていますが，このままだと下記のようなエラーが出てしまい，インクルードすることができません．正常にビルドできる場合，slnファイルのプロパティの「追加のインクルードディレクトリ」はどのように表示されますか？
![image](https://github.com/ut-issl/c2a-aobc/assets/131343764/51bbf81e-3ce1-47a4-9583-9d94a8677dc7)
![image](https://github.com/ut-issl/c2a-aobc/assets/131343764/85f2277f-21f8-49bd-b54d-249fd6ac684d)

**Reply by 200km** [reply]
@Sota2811 ありがとうございます。 @KotaToyoda さんとの議論から分かるように、`IncludeDirectories設定`が上手くいっていないようです。ISSL環境では、正しく`IncludeDirectories設定`ができているので画像で示してくれた部分でもエラーなどは発生せずにビルドできます。  

> slnファイルのプロパティの「追加のインクルードディレクトリ」はどのように表示されますか？

@conjikidow こちらについて、情報共有お願いします。

**Reply by 200km** [reply]
@Sota2811 こちらですが、ISSLとは別のユーザーから、そのユーザーの環境かではREADMEの通りの手順で動いたという報告をいただいております。  
お使いになられている環境特有の問題かもしれません。vMicro, Visual Studioのバージョン情報など教えていただけないでしょうか。

**Reply by Sota2811** [reply]
私たちの使用しているバージョンは
visual studio community 2019 16.11.29
visual micro for VS2019
です。

**Reply by 200km** [reply]
@Sota2811 現在、ISSLと他2組織の合計3組織がreadmeの手順で問題なく環境構築できているようです。ただ、それらはExampleを使わずにc2a-aobcを直接ビルドしています。（c2a-aobc直下のslnを開いてビルドする）  
それが問題なくできるか試してもらえないでしょうか？それがうまくいく場合、Exampleに関連する部分で説明やファイルにミスがある可能性があります。

**Reply by 200km** [reply]
@Sota2811 こちら、別組織の方にExampleベースでも手順確認していただき、問題なくビルドできるということの連絡をいただきました。  
手順のどこかで誤りがあるのかなと思っています。`example`という語を各プロジェクト名に置き換えていく部分などどのようにしたか、修正後のファイルを共有していただけないでしょうか。（slackで送っていただくのが良いかなと思います）

**Reply by Sota2811** [reply]
いつもお世話になっております．こちら，READMEの手順に加えて下記の手順でDTWIとDSPIをAdd libraryするとビルドが可能となりました．しかし，下記の写真のようにsrc_core, src_userがVisual Studio側でincludeできてないと判断しているようです．vMicroでビルドが成功しているということは，この赤波線の部分のエラーは無視してもよいのでしょうか？
![image](https://github.com/ut-issl/c2a-aobc/assets/131343764/9d5a93bd-443b-4180-bef4-1e02a1ea31bd)

**Reply by 200km** [reply]
> 下記の手順でDTWIとDSPIをAdd libraryするとビルドが可能となりました．

ビルドできるようになったようでよかったです。`この手順が必要になる可能性がある`ということで、readmeに注釈をつけておこうと思います。

> Visual Studio側でincludeできてない

Visual StudioのIntellisenceの設定の問題かなと思います。ビルドや実行に直接影響はないものだと思います。

**Comment by seki-hiro** [comment]
c2a-aobcはどのcommitのものを使われていますか？
[このcommit](https://github.com/ut-issl/c2a-aobc/commit/2d30f7296ac08fcad438053d09c8f0eb013a4dc2)で`C2A_AOBC_EXAMPLE.vcxproj`に修正が入ったので、それ以降のものを使うとうまくいくかもしれません。

**Reply by 200km** [reply]
そのcommitに関連する部分はまだreleaseされていないと思います。  
一般ユーザーはrelease版を使うことを想定しているので、submodule側のバージョン情報などもご確認ください。
